bindings/rust/aws-kms-tls-auth/src/provider.rs

```diff
@@ -0,0 +1,222 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    codec::EncodeValue,
+    identity::{ObfuscationKey, PskIdentity},
+    psk_from_material, KeyArn, KEY_ROTATION_PERIOD, PSK_SIZE,
+};
+use aws_sdk_kms::Client;
+use s2n_tls::{callbacks::ConnectionFuture, config::ConnectionInitializer};
+use std::{
+    pin::Pin,
+    sync::{Arc, RwLock},
+    time::Instant,
+};
+
+/// The `PskProvider` is used along with the [`PskReceiver`] to perform TLS
+/// 1.3 out-of-band PSK authentication, using PSK's generated from KMS.
+///
+/// This struct can be enabled on a config with [`s2n_tls::config::Builder::set_connection_initializer`].
+///
+/// The PSK is automatically rotated every 24 hours. Any errors in this rotation
+/// are logged through `tracing::error!`. Consider using something like
+/// [`tracing_subscriber`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/)
+/// to ensure visibility into these failures.
+#[derive(Debug, Clone)]
+pub struct PskProvider {
+    /// The KMS client
+    client: Client,
```

Comment:
We might want to explicitly call this "kms_client" or even just "kms", given the usual meaning of "client" for TLS contexts.